### ISSUE TITLE
Provide High-Level GPU abstraction

### ISSUE THREAD

**Issue by JayFoxRox** [XboxDev/nxdk#42]
I think we should implement some form of freeglut + subset of OpenGL 1.x (with NV specific extensions).

I did consider using MESA but I think it might be better to keep things minimalistic, as the primary use-case will be unit-tests for Xbox emulation. This means an OpenGL implementation should be rather flexible and easy to understand.

We probably also don't care too much about error-checking or 100% correct behaviour of OpenGL either.

I've started prototyping some (insanely ugly) OpenGL here https://github.com/JayFoxRox/nxdk/pull/3
Most of our existing code is also close to OpenGL already and could easily be hacked to have simple applications running.

**Comment by JayFoxRox** [comment]
Since this was created, a lot of things changed:

- [XGU is a higher-level-than-pbkit, low-level graphics API](https://github.com/dracc/nxdk/pull/4) (roughly 1 GPU command per function).
- XGUX is a higher-level-than-XGU, low-level graphics API (grouping GPU commands) that come with XGU. It also takes care to hide some GPU internals (example: converting floats to fixed-point and precomputing terms to be compatible with D3D or OpenGL parametrization).
- [xgu-gl is a very rough GL implementation on top of XGU](https://github.com/JayFoxRox/xgu-gl). It was [designed for my Neverball port](https://github.com/JayFoxRox/neverball/pull/1), but it's also being used in gish and some other projects of mine. It's not very generic (and it's a non-goal to be generic) and many features don't work as expected: dive into the codebase to adapt it for your project. It is a good base / snippet library for porting OpenGL (or even D3D) applications.

I also did look at Mesa (and even took some of its code), and I think it would have been a good choice. However, the direction that I took with xgu-gl is also fine for me and my projects.
I also agree with my initial statements that OpenGL should be usable for unit testing (also in tandem with low-level APIs); Mesa would have made this more difficult.

Most applications also only use a small subset of OpenGL, so implementing it yourself is very simple and allows for performance optimizations (also why xgu-gl isn't very generic).

I'll keep this issue opened, until XGU is officially "released", and until xgu-gl has better documentation (probably once my Neverball port is complete).
